File: library_net/src/main/java/com/hjc/library_net/converter/JsonResponseBodyConverter.java
package com.hjc.library_net.converter;

import androidx.annotation.NonNull;

import com.blankj.utilcode.util.LogUtils;
import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.JsonPrimitive;
import com.hjc.library_net.bean.BaseResponse;
import com.hjc.library_net.exception.ServerCode;

import java.io.IOException;
import java.lang.reflect.Type;

import okhttp3.ResponseBody;
import retrofit2.Converter;

/**
 * @Author: HJC
 * @Date: 2019/1/7 11:45
 * @Description: Gson对响应体进行处理
 */
public class JsonResponseBodyConverter<T> implements Converter<ResponseBody, T> {
    private final Gson mGson;
    private final Type mType;

    JsonResponseBodyConverter(Gson gson, Type type) {
        this.mGson = gson;
        this.mType = type;
    }

    @Override
    public T convert(@NonNull ResponseBody value) throws IOException {
        String json = value.string();
        LogUtils.e("response json: " + json);

        JsonObject jsonObject = JsonParser.parseString(json).getAsJsonObject();
        JsonPrimitive jsonPrimitive = jsonObject.getAsJsonPrimitive("code");

        if (jsonPrimitive != null) {
            String code = jsonPrimitive.getAsString();
            if (ServerCode.CODE_SUCCESS.equals(code)) {
                //返回结果正确,可以转换为对应的bean
                return mGson.fromJson(json, mType);
            } else {
                //接口请求失败,则直接转为BaseResponse
                return (T) mGson.fromJson(json, BaseResponse.class);
            }
        }
        return null;
    }
}

File: library-net/src/main/java/com/hjc/library_net/converter/JsonResponseBodyConverter.kt
package com.hjc.library_net.converter

import com.blankj.utilcode.util.LogUtils
import com.google.gson.Gson
import com.google.gson.JsonParser
import com.hjc.library_net.bean.BaseResponse
import com.hjc.library_net.exception.ServerCode
import okhttp3.ResponseBody
import retrofit2.Converter
import java.io.IOException
import java.lang.reflect.Type

/**
 * @Author: HJC
 * @Date: 2019/1/7 11:45
 * @Description: Gson对响应体进行处理
 */
@Suppress("UNCHECKED_CAST")
class JsonResponseBodyConverter<T> constructor(
    private val mGson: Gson,
    private val mType: Type
) : Converter<ResponseBody, T?> {

    @Throws(IOException::class)
    override fun convert(value: ResponseBody): T? {
        val json = value.string()
        LogUtils.e("response json: $json")

        val jsonObject = JsonParser.parseString(json).asJsonObject
        val jsonPrimitive = jsonObject.getAsJsonPrimitive("code")
        if (jsonPrimitive != null) {
            val code = jsonPrimitive.asInt
            return if (ServerCode.CODE_SUCCESS == code) {
                //返回结果正确,可以转换为对应的bean
                mGson.fromJson(json, mType)
            } else {
                //接口请求失败,则直接转为BaseResponse
                mGson.fromJson(json, BaseResponse::class.java) as T
            }
        }
        return null
    }
}